core/controller/src/main/scala/whisk/core/controller/Meta.scala

```diff
@@ -19,20 +19,127 @@ package whisk.core.controller
 import scala.concurrent.Future
 import scala.util.Failure
 import scala.util.Success
+import scala.util.Try
 
 import spray.http._
+import spray.http.HttpHeaders.RawHeader
+import spray.http.MediaTypes._
 import spray.http.StatusCodes._
 import spray.httpx.SprayJsonSupport._
 import spray.json._
 import spray.json.DefaultJsonProtocol._
 import spray.routing.Directives
+import spray.routing.RequestContext
 import whisk.common.TransactionId
 import whisk.core.controller.actions.PostActionActivation
 import whisk.core.database._
 import whisk.core.entity._
 import whisk.core.entity.types._
 import whisk.http.ErrorResponse.terminate
 import whisk.http.Messages
+import whisk.utils.JsHelpers._
+
+private case class Context(
+    method: HttpMethod,
+    headers: List[HttpHeader],
+    path: String,
+    query: Map[String, String],
+    body: Option[JsObject] = None) {
+    val requestParams = query.toJson.asJsObject.fields ++ { body.map(_.fields) getOrElse Map() }
+    val overrides = WhiskMetaApi.reservedProperties.intersect(requestParams.keySet)
+    def withBody(b: Option[JsObject]) = Context(method, headers, path, query, b)
+    def metadata(user: Option[Identity]) = {
+        WhiskMetaApi.metadata(method.value, headers.map(_.toString), path, user.map(_.namespace.asString))
+    }
+}
+
+object WhiskMetaApi extends Directives {
+    /** Reserved parameters that requests may no defined. */
+    val reservedProperties = Set(
+        "__ow_meta_verb",
+        "__ow_meta_headers",
+        "__ow_meta_path",
+        "__ow_meta_namespace")
+
+    /** Supported media types for action response. */
+    val mediaTranscoders = Map(
+        "html" -> resultAsHtml _,
+        "text" -> resultAsText _,
+        "js" -> resultAsJson _,
+        "http" -> resultAsHttp _)
+
+    val supportedMediaTypes = mediaTranscoders.keySet
+
+    def metadata(method: String, headers: List[String], path: String, namespace: Option[String]) = {
+        Map("__ow_meta_verb" -> method.toLowerCase.toJson,
+            "__ow_meta_headers" -> headers.toJson,
```

Comment:
I'm still wondering: What will this array look like now? Something like

```
["Authentication: Basic Foo", "Custom-Header": "bar"]
```
?

Would make sense to even pass the headers as a `Map` to make header handling closer to what *express* and other frameworks do in node.js for example?